docs: metrics (#3121)
bentoml/BentoML: docs/source/guides/metrics.rst

=======
Metrics
=======

Metrics are measurements of statistics about your service, which can provide information about the usage and performance of your bentos in production.

> BentoML allows users to define custom metrics with [#prometheus]_ to easily enable monitoring for their Bentos.

Reviewer: ```suggestion
BentoML allows users to define custom metrics with `Prometheus <https://prometheus.io/docs/introduction/overview/>`_ to easily enable monitoring for their Bentos.
```

 
This article will dive into how to add custom metrics to monitor your BentoService and how you can incorporate custom metrics into 
either a :ref:`concepts/runner:Custom Runner` or your :ref:`Service <concepts/service:Service and APIs>`.

> Make sure to have [#prometheus]_ installed and running before continuing.

Reviewer: ```suggestion
Having a `Prometheus server <https://prometheus.io/docs/prometheus/latest/getting_started/>` available will help visualize the examples in this guide.
```

.. note::

   This article assumes that you have a base understanding of a BentoService. If you
   are new to BentoML, please start with :ref:`the quickstart tutorial <tutorial:Tutorial: Intro to BentoML>`.

.. seealso::

   :ref:`Metrics API <reference/metrics>` for more information on ``bentoml.metrics``.

> Tracking model latency with Prometheus

Reviewer: ```suggestion
Using Metrics in a BentoService
```

>    :ref:`Metrics API <reference/metrics>` for more information on ``bentoml.metrics``.

Reviewer: ```suggestion
   All `metrics types <https://prometheus.io/docs/concepts/metric_types/>`_ supported by Prometheus are supported in BentoML. See :ref:`Metrics API <reference/metrics:Metrics API>` for more information on ``bentoml.metrics``.
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

We will build a custom histogram to track the latency of our :ref:`pretrained NLTK runner <concepts/runner:Custom Runner>`, a custom
counter to measure the total amount of time our endpoint is invoked.

.. note::

   The source code for this custom runner is :github:`available on GitHub <bentoml/BentoML/tree/main/examples/custom_runner/nltk_pretrained_model>`.

Initialize our metrics as follow:

.. literalinclude:: ./snippets/metrics/metric_defs.py
   :language: python
   :caption: `service.py`

``inference_duration`` is a :meth:`bentoml.metrics.Histogram`, which tracks how long it
takes for our model to run inference.

> The :attr:`bentoml.metrics.Histogram.buckets` is a exponential bucket with a factor of 1.5 and starts at 0.001.

Reviewer: ```suggestion
The :attr:`bentoml.metrics.Histogram.buckets` argument is used to determine the granularity of histogram tracking. The range of the buckets should cover the range of values the histogram should track. Number of buckets is positively correlated to the the granularity of tracking.
```

> ``num_invocation`` is a :meth:`bentoml.metrics.Counter`, which tracks the total number
> of invocation for any given endpoint.

Reviewer: ```suggestion
``num_invocation`` is a :meth:`bentoml.metrics.Counter`, which tracks the total number
of analysis by the polarity scores.
```

> ``num_invocation`` is a :meth:`bentoml.metrics.Counter`, which tracks the total number

Reviewer: ```suggestion
``polarity_counter`` is a :meth:`bentoml.metrics.Counter`, which tracks the total number
```

.. epigraph::

   :bdg-info:`Note:` This also applies to any other metric type, including :meth:`bentoml.metrics.Gauge` and :meth:`bentoml.metrics.Summary`.

Create our NLTK custom runner:

.. literalinclude:: ./snippets/metrics/runner_impl.py
   :language: python
   :caption: `service.py`

This runnable implementation creates a custom NLTK runner, that use the ``inference_duration``
histogram to track the latency of polatiry scores from a given sentence.

Initialize our NLTK runner, and add it to the service:

.. code-block:: python

   nltk_runner = t.cast(
      "RunnerImpl", bentoml.Runner(NLTKSentimentAnalysisRunnable, name="nltk_sentiment")
   )

>    svc = bentoml.Service("sentiment_analyzer", runners=[nltk_runner])

Reviewer: We no longer have to pass `metrics` to `bentoml.Service`?

   @svc.api(input=bentoml.io.Text(), output=bentoml.io.JSON())
   async def analysis(input_text: str) -> dict[str, bool]:

>        num_invocation.labels(endpoint="analysis").inc()
>        is_positive = await nltk_runner.is_positive.async_run(input_text)
>        return {"is_positive": is_positive}

Reviewer: ```suggestion
       is_positive = await nltk_runner.is_positive.async_run(input_text)
       polarity_counter.labels(polarity=is_positive).inc()
       return {"is_positive": is_positive}
```

> Our endpoint ``analysis`` uses the ``num_invocation`` counter to track the total number of
> invocation for ``analysis``.

Reviewer: ```suggestion
Our endpoint ``analysis`` uses the ``polarity_counter`` to track the total number of
invocation for ``analysis`` by polarity scores.
```

=======
Metrics
=======

Metrics are measurements of statistics about your service, which can provide information about the usage and performance of your bentos in production.

BentoML allows users to define custom metrics with `Prometheus <https://prometheus.io/>`_ to easily enable monitoring for their Bentos.
 

> This article will dive into how to add custom metrics to monitor your BentoService and how you can incorporate custom metrics into a :ref:`concepts/runner:Custom Runner`.

Reviewer: Can we expand the example to include API server as well?

Author: I think API server are similar to runner implementation. But I can add it in for the sake of completeness.

Make sure to have `Prometheus <https://prometheus.io/download/#prometheus>`_ installed and running before continuing.

Tracking model latency with Prometheus
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

We will build a custom histogram to track the latency of our :ref:`custom MNIST runner <concepts/runner:Custom Runner>` in this
tutorial.

.. note::

   The source code for this custom runner is :github:`available on GitHub <bentoml/BentoML/tree/main/examples/custom_model_runner>`.

Initialize our Histogram to track inference duration:

.. code-block:: python

   import bentoml

   inference_duration = bentoml.metrics.Histogram(
       name="inference_duration",
       documentation="Duration of inference",
       labelnames=["torch_version", "device_id"],
       buckets=exponential_buckets(0.001, 1.5, 10.0),
   )

This creates a :meth:`bentoml.metrics.Histogram`, which is a metric type that tracks the distribution of events with given bucket. The

> :attr:`bentoml.metrics.Histogram.buckets` is a exponential bucket with a factor of 1.5 and starts at 0.001.

Reviewer: Let's also callout that `Counter` and `Gauge` can be created here as well.

Follow with creating our custom MNIST runnable:

.. code-block:: python

   mnist_model = bentoml.pytorch.get("mnist_cnn:latest").
   _BuiltinRunnable = mnist_model.to_runnable()

   class MNISTRunnable(_BuiltinRunnable):
        def __init__(self):
            super().__init__()
            import torch

            print("Running on device:", self.device_id)
            self.torch_version = torch.__version__
            print("Running on torch version:", self.torch_version)

        @bentoml.Runnable.method(batchable=True, batch_dim=0)
        def __call__(self, input_arr: np.ndarray) -> np.ndarray:
            start = time.perf_counter()
            output = super().__call__(input_arr)
            inference_duration.labels(
                  torch_version=self.torch_version, device_id=self.device_id
            ).observe(time.perf_counter() - start)

>             return output.argmax(dim=1)

Reviewer: I think this example might be a bit too complex to demonstrate the metrics example. Can we maybe use a simpler example like `examples/custom_runner/nltk_pretrained_model`? Using the `nltk` example, we can also demonstrate the use of a `COUNTER` metric in the api_server when the return results is positive.

Author: Done.

=======
Metrics
=======

> Metrics are quantitative assessments for measuring and conducting assessments of your system's performance.
> Metrics are vital in providing reliable and accurate diagnoses to your BentoService in real-time scenarios.

Reviewer 2: ```suggestion
Metrics are measurements of statistics about your service, which can provide information about the usage and performance of your bentos in production.
```

> BentoML enables users to define custom metrics with `Prometheus <https://prometheus.io/>`_ to bootstrap and fast-track one's monitoring journey.

Reviewer 2: ```suggestion
BentoML allows users to define custom metrics with `Prometheus <https://prometheus.io/>`_ to easily enable monitoring for their bentos.
```

> This article will dive into how to add custom metrics to monitor your BentoService and a quick tutorial on how one can incorporate custom metrics to a :ref:`concepts/runner:Custom Runner`.

Reviewer 2: ```suggestion
This article will dive into how to add custom metrics to monitor your BentoService and how you can incorporate custom metrics into a :ref:`concepts/runner:Custom Runner`.
```

> Tracking model latency with Prometheus and :ref:`concepts/runner:Custom Runner`

Reviewer 2: ```suggestion
Tracking model latency with Prometheus and a :ref:`concepts/runner:Custom Runner`
```

-------------------------------------------------------------------------------

> We will build a custom histogram to track the latency of our custom Flax runner in this

Reviewer 2: I guess this is incomplete?

Author: Yes this is wip, I haven't push it up yet

tutorial.

.. note::

>    The example of this custom runner are :github:`available on GitHub <bentoml/BentoML/tree/main/examples/flax/seq2seq>`.

Reviewer 2: ```suggestion
   The source code for this custom runner is :github:`available on GitHub <bentoml/BentoML/tree/main/examples/flax/seq2seq>`.
```
